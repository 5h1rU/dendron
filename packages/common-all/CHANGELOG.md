# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.3.11](https://github.com/dendronhq/dendron/compare/v0.3.10...v0.3.11) (2020-07-13)

**Note:** Version bump only for package @dendronhq/common-all

## [0.3.10](https://github.com/dendronhq/dendron/compare/v0.1.6...v0.3.10) (2020-07-13)

### Features

- auto add nodes when deleted or created outside of dendron ([8c311bd](https://github.com/dendronhq/dendron/commit/8c311bda948a1d54088c49fd70eb65d24af5d68f))
- better schema suggestions ([03656bc](https://github.com/dendronhq/dendron/commit/03656bc007810457cb6846f0d6adacab4a7fbd3a))
- match namespace schemas ([7a67b8b](https://github.com/dendronhq/dendron/commit/7a67b8b2fb7caa1b97ee6d492d2801782abecdf6))
- show node descriptions ([aca86f2](https://github.com/dendronhq/dendron/commit/aca86f2a5fd6ee481f93553693a098db0e322890))
- surface unknown schemas ([d014965](https://github.com/dendronhq/dendron/commit/d0149652c985c69a4b2607984d578902820077f1))
- updated icons for schemas ([21804eb](https://github.com/dendronhq/dendron/commit/21804eba61c8dd49e499edd5d548d9d601224e8e))

## [0.3.9](https://github.com/dendronhq/dendron/compare/v0.3.8...v0.3.9) (2020-07-09)

### Features

- updated icons for schemas ([b7a2d8a](https://github.com/dendronhq/dendron/commit/b7a2d8aa517cf88d7a93d07cd2ef19305e48d069))

## [0.3.7](https://github.com/dendronhq/dendron/compare/v0.3.6...v0.3.7) (2020-07-08)

### Features

- better schema suggestions ([ad74bc0](https://github.com/dendronhq/dendron/commit/ad74bc009e1544319a49689394ab8d6b684f6578))
- show node descriptions ([e08fce9](https://github.com/dendronhq/dendron/commit/e08fce994153e28fe504b85e6d9bc1f5fdd93e20))

## [0.3.6](https://github.com/dendronhq/dendron/compare/v0.3.5...v0.3.6) (2020-07-07)

### Features

- auto add nodes when deleted or created outside of dendron ([a7e1ac9](https://github.com/dendronhq/dendron/commit/a7e1ac9b8a4f7f0592ab1b9f86a7a40182693a73))

# [0.3.0](https://github.com/dendronhq/dendron/compare/v0.2.20...v0.3.0) (2020-07-05)

### Features

- match namespace schemas ([1a960bf](https://github.com/dendronhq/dendron/commit/1a960bf26f8984e541b3eb118f60bdc09d8250fe))
- surface unknown schemas ([9bf4d0e](https://github.com/dendronhq/dendron/commit/9bf4d0e61cce2f76bddae1f686f29474201466cb))
