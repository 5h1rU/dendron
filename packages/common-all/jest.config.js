const jestConfig = require("../../jest.config");

console.log(jestConfig);
module.exports = {
  ...jestConfig
};
