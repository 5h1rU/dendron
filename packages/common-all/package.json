{
  "name": "@dendronhq/common-all",
  "version": "0.1.6",
  "description": "common-all",
  "license": "MIT",
  "repository": "https://github.com/dendronhq/dendron",
  "author": {
    "name": "Kevin Lin",
    "email": "kevin@thence.io",
    "url": "https://kevinslin.com",
    "organization": true
  },
  "keywords": [
    "common"
  ],
  "files": [
    "lib"
  ],
  "main": "lib/index",
  "types": "lib/index",
  "scripts": {
    "clean": "rimraf lib && rimraf coverage",
    "format": "",
    "lint": "",
    "prepublishOnly": "npm run build",
    "prebuild": "npm run clean && npm run format && npm run lint && echo Using TypeScript && tsc --version",
    "build": "tsc --pretty",
    "test": "jest",
    "coverage": "jest --coverage",
    "watch": "npm run build -- --watch",
    "test:watch": "jest --watch"
  },
  "dependencies": {
    "dropbox": "^4.0.30",
    "isomorphic-fetch": "^2.2.1",
    "lodash": "^4.17.15",
    "pino": "^6.3.1",
    "uuid": "^8.1.0",
    "yamljs": "^0.3.0"
  },
  "devDependencies": {
    "@types/jest": "^23.3.3",
    "@types/lodash": "^4.14.152",
    "@types/node": "^10.11.4",
    "@types/pino": "^6.0.1",
    "@types/uuid": "^8.0.0",
    "@types/yamljs": "^0.2.31",
    "coveralls": "^3.0.2",
    "jest": "^23.6.0",
    "prettier": "^1.14.3",
    "rimraf": "^2.6.2",
    "ts-jest": "^23.10.3",
    "ts-node": "^7.0.1",
    "tslint": "^5.11.0",
    "tslint-config-prettier": "^1.15.0",
    "typescript": "^3.1.1"
  },
  "engines": {
    "node": ">=10.0.0"
  },
  "jest": {
    "preset": "ts-jest"
  },
  "gitHead": "403eb012f8b34bc115dd7dd86f80c76b39329f3b"
}
