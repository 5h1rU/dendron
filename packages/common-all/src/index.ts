export * from "./types";
export * from "./store";
export * from "./node";
export * from "./helpers";
