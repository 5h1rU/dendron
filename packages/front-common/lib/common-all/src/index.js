export * from "./node";
export * from "./helpers";
export * from "./integrations/dropbox";
//# sourceMappingURL=index.js.map