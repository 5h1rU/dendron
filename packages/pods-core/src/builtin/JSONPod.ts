import { Note, NoteRawProps } from "@dendronhq/common-all";
import fs from "fs-extra";
import _ from "lodash";
import { URI } from "vscode-uri";
import {
  ExportConfig,
  ExportPod,
  ExportPodBaseV2,
  ExportPodOpts,
  ImportConfig,
  ImportPodBaseV2,
  ImportPodOpts,
  PodConfigEntry,
} from "../base";

const ID = "dendron.pod.json";

export type ImportPodConfig = ImportConfig & {
  concatenate: boolean;
  destName?: string;
};

export class JSONImportPod extends ImportPodBaseV2<ImportPodConfig> {
  static id: string = ID;
  static description: string = "import to json";

  static config = (): PodConfigEntry[] => {
    return [
      {
        key: "src",
        description: "where will notes be imported from",
        type: "string",
      },
      {
        key: "concatenate",
        description:
          "concatenate all entries into one note? if set to true, need to set `destName`",
        type: "boolean",
        default: false,
      },
      {
        key: "destName",
        description:
          "if `concatenate: true`, specify name of concatenated note",
        type: "string",
      },
    ];
  };

  async plant(opts: ImportPodOpts<ImportPodConfig>): Promise<void> {
    const cleanConfig = this.cleanConfig(opts.config);
    await this.prepare(opts);
    await this.execute({ ...opts.config, ...cleanConfig });
  }

  async execute(opts: { src: URI } & Omit<ImportPodConfig, "src">) {
    const ctx = "JSONPod";
    this.L.info({ ctx, opts, msg: "enter" });
    const { src, destName, concatenate } = opts;
    const entries = fs.readJSONSync(src.fsPath);
    const notes = await this._entries2Notes(entries, { destName, concatenate });
    return Promise.all(
      _.map(notes, (n) => this.engine.write(n, { newNode: true }))
    );
  }

  async _entries2Notes(
    entries: Partial<NoteRawProps>[],
    opts: Pick<ImportPodConfig, "concatenate" | "destName">
  ): Promise<Note[]> {
    const notes = _.map(entries, (ent) => {
      if (!ent.fname) {
        throw Error("fname not defined");
      }
      let fname = ent.fname;
      return new Note({ ...ent, fname, parent: null, children: [] });
    });
    if (opts.concatenate) {
      if (!opts.destName) {
        throw Error(
          "destname needs to be specified if concatenate is set to true"
        );
      }
      const acc: string[] = [""];
      _.forEach(notes, (n) => {
        acc.push(`# ${n.fname}`);
        acc.push(n.renderBody());
        acc.push("---");
      });
      return [new Note({ fname: opts.destName, body: acc.join("\n") })];
    } else {
      return notes;
    }
  }
}

export class JSONExportPod extends ExportPodBaseV2
  implements ExportPod<ExportConfig> {
  static id: string = ID;
  static description: string = "export to json";

  static config = (): PodConfigEntry[] => {
    return [
      {
        key: "dest",
        description: "where will output be stored",
        type: "string",
      },
    ];
  };

  async plant(opts: ExportPodOpts<ExportConfig>): Promise<void> {
    return new Promise(async (resolve) => {
      await this.initEngine();
      const cleanConfig = this.cleanConfig(opts.config);
      const payload = this.prepareForExport(opts);
      const destPath = cleanConfig.dest.fsPath;
      fs.writeJSONSync(destPath, payload, { encoding: "utf8" });
      resolve();
    });
  }
}
