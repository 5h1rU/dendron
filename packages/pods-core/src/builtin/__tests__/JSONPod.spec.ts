import { DNodeRawProps, Note, NoteRawProps } from "@dendronhq/common-all";
import {
  EngineTestUtils,
  FileTestUtils,
  node2MdFile,
} from "@dendronhq/common-server";
import fs from "fs-extra";
import _ from "lodash";
import path from "path";
import { URI } from "vscode-uri";
import { ExportPod } from "../JSONPod";

const createNotes = (vaultPath: string, notes: Partial<NoteRawProps>[]) => {
  node2MdFile(new Note({ fname: "root", id: "root", title: "root" }), {
    root: vaultPath,
  });
  notes.map((n) => {
    // @ts-ignore
    node2MdFile(new Note(n), {
      root: vaultPath,
    });
  });
};

const assertNodeMeta = (opts: {
  expect: jest.Expect;
  payload: DNodeRawProps[];
  fields: string[];
  expected: Partial<DNodeRawProps>[];
}) => {
  const { expect, payload, fields, expected } = opts;
  expect(_.sortBy(_.map(payload, (ent) => _.pick(ent, fields)))).toEqual(
    expected
  );
};

const assertNodeBody = (opts: {
  expect: jest.Expect;
  payload: DNodeRawProps[];
  expected: { fname: string; body: string }[];
}) => {
  const { expect, payload, expected } = opts;
  expect(
    _.sortBy(
      _.map(payload, (ent) => {
        const { body, fname } = _.pick(ent, ["body", "fname"]);
        return {
          fname,
          body: _.trim(body),
        };
      })
    )
  ).toEqual(expected);
};

function setup(opts: { notes: Partial<NoteRawProps>[] }) {
  return EngineTestUtils.setupStoreDir({
    copyFixtures: false,
    initDirCb: (dirPath: string) => {
      createNotes(dirPath, opts.notes);
    },
  });
}

describe("ExportPod", () => {
  let root: string;

  const createNotes = (): Partial<NoteRawProps>[] => [
    { fname: "foo", body: "foo body" },
    { fname: "bar", body: "bar body" },
  ];

  test("basic", async () => {
    root = await setup({ notes: createNotes() });
    const pod = new ExportPod({ roots: [root] });
    const mode = "notes";
    const metaOnly = false;
    const destDir = FileTestUtils.tmpDir().name;
    const destPath = path.join(destDir, "export.json");
    const config = { dest: URI.file(destPath) };
    await pod.plant({ mode, metaOnly, config });
    const payload = fs.readJSONSync(destPath) as NoteRawProps[];
    expect(payload).toMatchSnapshot();
    assertNodeMeta({
      expect,
      payload,
      fields: ["fname"],
      expected: [
        {
          fname: "root",
        },
        { fname: "bar" },
        { fname: "foo" },
      ],
    });
    assertNodeBody({
      expect,
      payload,
      expected: [
        {
          fname: "root",
          body: "",
        },
        { fname: "bar", body: "bar body" },
        { fname: "foo", body: "foo body" },
      ],
    });
  });

  test("basic no body", async () => {
    root = await setup({ notes: createNotes() });
    const pod = new ExportPod({ roots: [root] });
    const mode = "notes";
    const metaOnly = true;
    const destDir = FileTestUtils.tmpDir().name;
    const destPath = path.join(destDir, "export.json");
    const config = { dest: URI.file(destPath) };
    await pod.plant({ mode, metaOnly, config });
    const payload = fs.readJSONSync(destPath) as NoteRawProps[];
    expect(payload).toMatchSnapshot();
    assertNodeMeta({
      expect,
      payload,
      fields: ["fname"],
      expected: [
        {
          fname: "root",
        },
        { fname: "bar" },
        { fname: "foo" },
      ],
    });
    assertNodeBody({
      expect,
      payload,
      expected: [
        {
          fname: "root",
          body: "",
        },
        { fname: "bar", body: "" },
        { fname: "foo", body: "" },
      ],
    });
  });
});
