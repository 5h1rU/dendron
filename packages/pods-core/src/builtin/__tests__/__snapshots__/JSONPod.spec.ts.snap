// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSONExportPod write config 1`] = `
"# description: where will output be stored
# type: string
dest: TODO"
`;

exports[`JSONImportPod concatenate 1`] = `
"---
id: ad9f3347-7adf-49b0-97e8-acede4b93c5a
title: Results
desc: ''
updated: 1600110905047
created: 1600110905047
stub: false
---

# [[foo]]
foo body
---
# [[bar]]
bar body
---
"
`;
